File: backend-nestjs-feature/src/features/export/export.controller.ts
import { Controller, Post, Body, Res, UseGuards } from '@nestjs/common';
import { Response } from 'express';
import { ExportService } from './export.service';
import { AuthGuard } from '@nestjs/passport';

@UseGuards(AuthGuard('jwt'))
@Controller('export')
export class ExportController {
  constructor(private readonly exportService: ExportService) {}

  @Post('csv')
  async exportCsv(@Body() data: any[], @Res() res: Response) {
    const buffer = await this.exportService.exportToCsv(data);
    res.set({
      'Content-Type': 'text/csv',
      'Content-Disposition': `attachment; filename="export_${Date.now()}.csv"`,
      'Content-Length': buffer.length,
    });
    res.end(buffer);
  }

  @Post('excel')
  async exportExcel(@Body() data: any[], @Res() res: Response) {
    const buffer = await this.exportService.exportToExcel(data);
    res.set({
      'Content-Type': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
      'Content-Disposition': `attachment; filename="export_${Date.now()}.xlsx"`,
      'Content-Length': buffer.length,
    });
    res.end(buffer);
  }
}

File: backend/spring/feature-architecture/full/src/main/java/com/apptemplate/api/features/export/ExportController.java
package com.apptemplate.api.features.export;

import lombok.RequiredArgsConstructor;
import org.springframework.core.io.InputStreamResource;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/export")
@RequiredArgsConstructor
public class ExportController {

    private final ExportService exportService;

    @PostMapping("/csv")
    public ResponseEntity<Resource> exportCsv(@RequestBody List<Map<String, Object>> data) {
        InputStreamResource file = new InputStreamResource(exportService.exportToCsv(data));
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=export.csv")
                .contentType(MediaType.parseMediaType("text/csv"))
                .body(file);
    }

    @PostMapping("/excel")
    public ResponseEntity<Resource> exportExcel(@RequestBody List<Map<String, Object>> data) {
        InputStreamResource file = new InputStreamResource(exportService.exportToExcel(data));
        return ResponseEntity.ok()
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=export.xlsx")
                .contentType(MediaType.parseMediaType("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"))
                .body(file);
    }
}
